src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -2236,6 +2231,14 @@ private bool IsTypeDeclarationStart()
                 case SyntaxKind.ExplicitKeyword or SyntaxKind.ImplicitKeyword
                     when this.PeekToken(1).ContextualKind is SyntaxKind.ExtensionKeyword:
 
+                    // If we have `implicit extension E` then this is definitely an extension as it could not be parsed
+                    // as anything else.
+                    if (IsTrueIdentifier(this.PeekToken(2)))
+                        return true;
+
+                    // we have `implicit extension ...` technically this could be something like `implicit
+                    // extension.operator X(` (an implicit conversion with a explicit interface impl name). For compat,
+                    // determine the parsing strategy based on the lang version.
                     return IsFeatureEnabled(MessageID.IDS_FeatureExtensions);
```

Comment:
This should probably use `IsCompatBreakingFeatureEnabled` to signal that we're taking an intentional break.
In terms of documentation, I think it'll be covered when we add the logic to disallow types named "extension" (which is tracked in the test plan).

Reply:
done!